### PR TITLE
Validate file and handle exceptions for weights_only unpickler

### LINKED ISSUE #127525: Multiple unhandled exceptions in weights_only unpickler
### 🐛 Describe the bug

Hi!

During model loading with Python API `torch.load(filename, weights_only=True)` a many parsing errors may take place. Some errors are handled with catching exceptions from Unpickler module, but many of them are unhandled. Loading of corrupted or inconsistent file may result in IndexError and EOFError exceptions:
```
 === Uncaught Python exception: ===
IndexError: list index out of range
Traceback (most recent call last):
  File "/load_weights_fuzz.py", line 42, in TestOneInput
    torch.load(filename, weights_only=True)
  File "/usr/local/lib/python3.9/dist-packages/torch/serialization.py", line 1037, in load
    return _legacy_load(opened_file, map_location, _weights_only_unpickler, **pickle_load_args)
  File "/usr/local/lib/python3.9/dist-packages/torch/serialization.py", line 1262, in _legacy_load
    magic_number = pickle_module.load(f, **pickle_load_args)
  File "/usr/local/lib/python3.9/dist-packages/torch/_weights_only_unpickler.py", line 306, in load
    return Unpickler(file, encoding=encoding).load()
  File "/usr/local/lib/python3.9/dist-packages/torch/_weights_only_unpickler.py", line 220, in load
    self.stack[-3:] = [(self.stack[-3], self.stack[-2], self.stack[-1])]
IndexError: list index out of range
```

Some of the seeds to trigger bugs: [crash-1.zip](https://github.com/pytorch/pytorch/files/15499489/crash-1.zip), [crash-2.zip](https://github.com/pytorch/pytorch/files/15499490/crash-2.zip)

### Versions

PyTorch version: 2.3.0, current main

OS: Ubuntu 20.04.6 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.2) 9.4.0
Clang version: 14.0.6
CMake version: version 3.29.2
Libc version: glibc-2.31

Python version: 3.9.5 (default, Nov 23 2021, 15:27:38) [GCC 9.3.0] (64-bit runtime)
Python platform: Linux-5.4.0-174-generic-x86_64-with-glibc2.31

### PR DESCRIPTION
**Pull request by apach301**
Fixing a lot of unhandled exceptions raised from weights_only model loading:

 - Validate file size to be able to contain at least magic number and protocol version
 - Also catch all exceptions from unpickler due to inconsistent operators interpretation.

Fixes #127525

### CODE REVIEW

File: torch/serialization.py [hunk 1]
```diff
@@ -1269,6 +1269,16 @@ def persistent_load(saved_id):
             f"Received object of type \"{type(f)}\". Please update to Python 3.8.2 or newer to restore this "
             "functionality.")
 
+    start = f.tell()
+    f.seek(0, 2)
+    file_size = f.tell()
+    # Check that file is long enough to contain magic number and
+    # protocol version: 10 bytes for MAGIC_NUMBER, 2 bytes for
+    # PROTOCOL_VERSION, 8 bytes for unpickling operators.
```
Comment (mikaylagawarecki) [inline torch/serialization.py:1277]: Could you elaborate on why 8 bytes?

**Review by mikaylagawarecki — COMMENTED** [review]
Thanks, could you add a test please

File: torch/serialization.py (hunk 1 above)
Comment (apach301) [reply torch/serialization.py:1277]: Valid test that I tested: 2 bytes for PROTO (start unpickling), 2 bytes for LONG1 and n_bytes, [10 bytes for magic number], 1 byte for STOP, 2 bytes for PROTO, 1 byte for BININT2, [2 bytes for protocol].

Also after protocol there should be 1 byte for STOP (thus, 9 bytes of system info), and after that _sys_info is also expected to be in file, but the main idea of this check is to simply validate that file is long enough to contain constant values that we try to read and check.

**Comment by apach301** [comment]
@mikaylagawarecki Finally fixed a test, how could I re-run CI?

**Pull request closed without merging**